src/python_inspector/setup_py_live_eval.py

```diff
@@ -0,0 +1,188 @@
+    """Parse set from comma separated string."""
+    string = string.strip()
+    if string:
+        return set(string.split(","))
+    else:
+        return set()
+
+
+def minver_error(pkg_name):
+    """Report error about missing minimum version constraint and exit."""
+    print(
+        'ERROR: specify minimal version of "{0}" using ' '">=" or "=="'.format(pkg_name),
+        file=sys.stderr,
+    )
+    sys.exit(1)
+
+
+def build_pkg_name(pkg):
+    """Build package name, including extras if present."""
+    if pkg.extras:
+        return "{0}[{1}]".format(pkg.project_name, ",".join(sorted(pkg.extras)))
+    return pkg.project_name
+
+
+def parse_pip_file(path):
+    """Parse pip requirements file."""
+    # requirement lines sorted by importance
+    # also collect other pip commands
+    rdev = {}
+    rnormal = []
+    stuff = []
+
+    try:
+        with open(path) as f:
+            for line in f:
+                line = line.strip()
+
+                # see https://pip.readthedocs.io/en/1.1/requirements.html
+                if line.startswith("-e"):
+                    # devel requirement
+                    splitted = line.split("#egg=")
+                    rdev[splitted[1].lower()] = line
+
+                elif line.startswith("-r"):
+                    # recursive file command
+                    splitted = re.split("-r\\s+", line)
+                    subrdev, subrnormal, substuff = parse_pip_file(
+                        os.path.join(os.path.dirname(path), splitted[1])
+                    )
+                    for k, v in subrdev.items():
+                        if k not in rdev:
+                            rdev[k] = v
+                    rnormal.extend(subrnormal)
+                elif line.startswith("-"):
+                    # another special command we don't recognize
+                    stuff.append(line)
+                else:
+                    # ordinary requirement, similarly to them used in setup.py
+                    rnormal.append(line)
+    except IOError:
+        print('Warning: could not parse requirements file "{0}"!'.format(path), file=sys.stderr)
+
+    return rdev, rnormal, stuff
+
+
+def iter_requirements(level, extras, setup_file):
+    """Iterate over requirements."""
+    from pathlib import Path
+
+    setup_file = str(Path(setup_file).absolute())
+    result = dict()
+    requires = []
+    stuff = []
+    cd = os.getcwd()
+    os.chdir(os.path.dirname(setup_file))
+    install_requires = []
+    requires_extras = {}
+    # change directory to setup.py path
+    with mock.patch.object(setuptools, "setup") as mock_setup:
+        sys.path.append(os.path.dirname(setup_file))
+        g = {"__file__": setup_file, "__name__": "__main__"}
+        with open(setup_file) as sf:
+            exec(sf.read(), g)
+        sys.path.pop()
+        assert g["setup"]  # silence warning about unused imports
+    # called arguments are in `mock_setup.call_args`
+    os.chdir(cd)
```

Comment:
This should be in a context manager as we cannot change the cwd forever IMHO. You can use this `commoncode.command.pushd`https://github.com/nexB/commoncode/blob/9131627677d3ef171ddc472991a5c4d4a3431ee3/src/commoncode/command.py#L246